### DISCUSSION TITLE
Incremental deployments

### DISCUSSION THREAD

**Discussion by martinsson** [serverless/serverless#10900] [Q&A]
I find deployments to be unnecessarily slow which greatly limits our ability to deploy often and quickly

Currently if a single lambda in a service changes the whole service is redeployed. This includes
- re-uploading all functions (even though their checksum did not change)
- applying an update to every function (which is long due to aws SAM from what I understand)

It'd be great if functions that did not change were not re-uploaded and it'd be great if there was some sort
of diff algorithm to the serverless descriptor file to exclude some things from the SAM update. For some-one 
not familiar with the underlying code this seems like a not-so-hard feature to implement. It also seems like a total killer feature so
I'm puzzled to why this is not implemented or even visible on the roadmap. Seed has this since quite some time for instance.

There's also an extension to this idea of incremental deploys : Now that we're in the world of immutability, we have git, we have docker-hub and all sorts of tools that avoids re-building or re-storing the same thing. So why don't we a registry for lambdas? For instance once a compiled lambda binary has been uploaded to S3 for one stage why does it have to be re-uploaded to another stage?

So my question is, is this planned for any near future or is this subject to important technical constraints?
Also isn't everyone impatient for this feature?

**Comment by ennioVisco** [comment]
I totally agree with @martinsson, the deployment time is very annoying. 
In a project I'm working one I have approx. 40 lambdas and it's already taking ~3mins for every single deployment (only to run `sls deploy`), even when just changing 1 line of code.

Couldn't serverless just cache previous builds and avoid re-uploading when the hash of the file is the same??

This would probably bring the deployment time down by 90% or similar in average deployments...

**Reply by dservidie** [reply]
You can also do:

```sls deploy function -f functionToDeploy```

And deploy just one function. It's faster, but for sure this should be even faster. Also, plugins for testing locally are precarious. There should be a better way.
